ibm/resource_ibm_firewall_shared.go

```diff
@@ -0,0 +1,224 @@
+	targetItems := []datatypes.Product_Item{}
+	for _, item := range productItems {
+		if *item.KeyName == keyName {
+			targetItems = append(targetItems, item)
+		}
+	}
+
+	if len(targetItems) == 0 {
+		return fmt.Errorf("No product items matching %s could be found", keyName)
+	}
+
+	if guestType == "virtual machine" {
+		productOrderContainer := datatypes.Container_Product_Order_Network_Protection_Firewall{
+			Container_Product_Order: datatypes.Container_Product_Order{
+				PackageId: pkg.Id,
+				Prices: []datatypes.Product_Item_Price{
+					{
+						Id: targetItems[0].Prices[0].Id,
+					},
+				},
+				Quantity: sl.Int(1),
+				VirtualGuests: []datatypes.Virtual_Guest{{
+					Id: sl.Int(machineId),
+				},
+				},
+			},
+		}
+		receipt, err := services.GetProductOrderService(sess.SetRetries(0)).PlaceOrder(&productOrderContainer, sl.Bool(false))
+		log.Print("receipt for order placed")
+		log.Print(receipt)
+		if err != nil {
+			return fmt.Errorf("Error during creation of hardware firewall: %s", err)
+		}
+
+	}
+	if guestType == "baremetal" {
+		productOrderContainer := datatypes.Container_Product_Order_Network_Protection_Firewall{
+			Container_Product_Order: datatypes.Container_Product_Order{
+				PackageId: pkg.Id,
+				Prices: []datatypes.Product_Item_Price{
+					{
+						Id: targetItems[0].Prices[0].Id,
+					},
+				},
+				Quantity: sl.Int(1),
+				Hardware: []datatypes.Hardware{{
+					Id: sl.Int(machineId),
+				},
+				},
+			},
+		}
+		receipt, err := services.GetProductOrderService(sess.SetRetries(0)).PlaceOrder(&productOrderContainer, sl.Bool(false))
+		log.Print("receipt for order placed")
+		log.Print(receipt)
+		if err != nil {
+			return fmt.Errorf("Error during creation of hardware firewall: %s", err)
+		}
+
+	}
+	log.Println("[INFO] Creating hardware firewall shared")
+
+	d.Set("firewall_type", keyName)
+	d.Set("guest_id", machineId)
+	d.Set("guest_type", guestType)
+
+	log.Printf("[INFO] Wait one minute before fetching the firewall/device.")
+	time.Sleep(time.Second * 30)
+
+	return resourceIBMFirewallSharedRead(d, meta)
+}
+
+func resourceIBMFirewallSharedRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	macId := (d.Get("guest_id").(int))
+
+	guestType := (d.Get("guest_type").(string))
+
+	masked := "firewallServiceComponent.id"
+
+	fservice := services.GetNetworkComponentFirewallService(sess)
+
+	if guestType == "virtual machine" {
+		service := services.GetVirtualGuestService(meta.(ClientSession).SoftLayerSession())
+		result, err := service.Id(macId).Mask(masked).GetObject()
+
+		if err != nil {
+			return fmt.Errorf("Error retrieving firewall information: %s", err)
+		}
+
+		if result.FirewallServiceComponent == nil {
+			return fmt.Errorf("Error retrieving firewall information.This resource has already been canceled.")
+		}
+		idd := *result.FirewallServiceComponent.Id
+
+		d.SetId(fmt.Sprintf("%d", idd))
```

Comment:
Set id as combination of firewallid/<id of virtual /hardware>

In import scenarios user can't use d.Get() since we will not have configuration file...
So set the id d.SetId(fmt.Sprintf("%d/%d", idd, macId)) and move the logic of setting ID to Create Method..